### PR TITLE
Adds expr functions for KV, KVV and Vars manipulation

### PR DESCRIPTION
**Pull request by vicpatel**
Adds expr functions for KV, KVV, Vars manipulation

- Adds/Refactors methods for Set, Merge, Filter, Delete
- Renames expr/Vars.Merge to MustMerge, updates its usage
- Appends kvFunctions to parser in pkg/expr
- Update/Fixes tests and example testable

Please take a look.

### CODE REVIEW

File: pkg/expr/expr_types.go [hunk 1]
```diff
@@ -23,6 +23,26 @@ type (
 		Type
 		ResolveTypes(func(string) Type) error
 	}
+
+	iterator interface {
+		Each(func(string, interface{}) error) error
+	}
+
+	setter interface {
```
Comment (darh) [inline pkg/expr/expr_types.go]: could an existing `FieldAssigner` interface be used instead of this, new one?

File: pkg/expr/expr_types.go [hunk 2]
```diff
@@ -23,6 +23,26 @@ type (
 		Type
 		ResolveTypes(func(string) Type) error
 	}
+
+	iterator interface {
```
Comment (darh) [inline pkg/expr/expr_types.go]: could an existing `Iterator` be used instead of this, new one?

**Comment by darh** [comment]
btw, regarding interface `filter`

https://golang.org/doc/effective_go#interface-names

and you would not be the first one to go with `filterer`, no matter how quirky it sounds
https://pkg.go.dev/github.com/msample/logfu

:)

File: pkg/expr/expr_types.go (hunk 1 above)
Comment (vicpatel) [reply pkg/expr/expr_types.go]: Yes

File: pkg/expr/values.go [hunk 3]
```diff
@@ -30,6 +30,7 @@ type (
 
 	FieldAssigner interface {
 		AssignFieldValue(string, TypedValue) error
+		Set(string, interface{}) error
```
Comment (darh) [inline pkg/expr/values.go]: Why? :) AssignFieldValue does exact same thing already.

File: pkg/expr/expr_types.go [hunk 4]
```diff
@@ -404,6 +432,85 @@ func CastToKV(val interface{}) (out map[string]string, err error) {
 	}
 }
 
+func (t *KV) Each(fn func(k string, v TypedValue) error) (err error) {
+	if t == nil || t.value == nil {
+		return
+	}
+
+	for k, v := range t.value {
+		if err = fn(k, Must(NewString(v))); err != nil {
```
Comment (darh) [inline pkg/expr/expr_types.go]: Must will panic. You use it when it's ok to panic (and it's not in most cases).
You return error from Each anyway, so better to try and do NewString, and return if it yields an error.

File: pkg/expr/expr_types.go [hunk 5]
```diff
@@ -404,6 +432,85 @@ func CastToKV(val interface{}) (out map[string]string, err error) {
...
+	if t.value == nil {
+		t.value = make(map[string]string)
+	}
+
+	nn = append([]Iterator{t}, nn...)
+
+	for _, i := range nn {
+		if err = i.Each(func(k string, v TypedValue) error {
+			t.value[k] = cast.ToString(v.Get())
+			return err
```
Comment (darh) [inline pkg/expr/expr_types.go]: What error are you returning here exactly?

File: pkg/expr/expr_types.go [hunk 6]
```diff
@@ -404,6 +432,85 @@ func CastToKV(val interface{}) (out map[string]string, err error) {
...
+func (t *KV) Merge(nn ...Iterator) (err error) {
+	if t.value == nil {
+		t.value = make(map[string]string)
+	}
+
+	nn = append([]Iterator{t}, nn...)
+
+	for _, i := range nn {
+		if err = i.Each(func(k string, v TypedValue) error {
+			t.value[k] = cast.ToString(v.Get())
```
Comment (darh) [inline pkg/expr/expr_types.go]: you can use `cast.ToStringE` to get the fn variant that does not suppress your errors

File: pkg/expr/expr_types.go [hunk 7]
```diff
@@ -404,6 +432,85 @@ func CastToKV(val interface{}) (out map[string]string, err error) {
...
+// Merge combines the given KVs into KV
+func (t *KV) Merge(nn ...Iterator) (err error) {
+	if t.value == nil {
+		t.value = make(map[string]string)
+	}
+
+	nn = append([]Iterator{t}, nn...)
+
+	for _, i := range nn {
+		if err = i.Each(func(k string, v TypedValue) error {
```
Comment (darh) [inline pkg/expr/expr_types.go]: please keep _lambdas_ out of ifs

File: pkg/expr/vars.go [hunk 8]
```diff
@@ -409,3 +410,36 @@ func CastToVars(val interface{}) (out map[string]TypedValue, err error) {
...
+// Filter take keys returns KV with only those key value pair
+func (t *Vars) Filter(keys ...string) (err error) {
+	if t.value == nil {
+		return
+	}
+
+	// Delete key from t.value if it is not from keys
+	for k := range t.value {
+		if !contains(keys, k) {
+			delete(t.value, k)
```
Comment (darh) [inline pkg/expr/vars.go]: no need to pre-check for key for delete fn.

File: pkg/expr/expr_types.yaml [hunk 9]
```diff
@@ -37,6 +37,9 @@ types:
     as: 'string'
     default: '""'
 
+  StringSlice:
```
Comment (darh) [inline pkg/expr/expr_types.yaml]: do we need this?

**Review by darh — CHANGES_REQUESTED** [review]

File: pkg/expr/expr_types.yaml (hunk 9 above)
Comment (vicpatel) [reply pkg/expr/expr_types.yaml]: No

File: pkg/expr/expr_types.go (hunk 5 above)
Comment (vicpatel) [reply pkg/expr/expr_types.go]: Fixed it

**Review by darh — CHANGES_REQUESTED** [review]
these new functions still need to be aded to `expr` pkg, see existing function mappings (via `gval.Function`) in `pkg/expr/func_*.go`

File: pkg/expr/expr_types_test.go [hunk 10]
```diff
@@ -8,6 +8,149 @@ import (
...
+			"kvv":   &KVV{},
+			"key":   "foo",
+			"value": Must(NewString("bar")),
+		}
+	)
+
+	eval(`set(kvv, key, value)`, p)
+
+	// output:
+	// <nil>
```
Comment (darh) [inline pkg/expr/expr_types_test.go]: Is this correct? Should test return <nil>? Or modified input?
So, is original input mutated here or does it stay unchanged after call? Would be very nice if we could prove that it's unchanged with the test?

File: pkg/expr/expr_types_test.go [hunk 11]
```diff
@@ -8,6 +8,149 @@ import (
...
+			"kv":    &KV{},
+			"key":   "k1",
+			"value": Must(NewString("v11")),
+		}
+	)
+
+	eval(`set(kv, key, value)`, p)
+
+	// output:
+	// <nil>
```
Comment (darh) [inline pkg/expr/expr_types_test.go]: same as for Example_set_kvv

File: pkg/expr/vars_test.go [hunk 12]
```diff
@@ -8,6 +8,69 @@ import (
...
+			"vars":  &Vars{},
+			"key":   "foo",
+			"value": &String{value: "foo"},
+		}
+	)
+
+	eval(`set(vars, key, value)`, p)
+
+	// output:
+	// <nil>
```
Comment (darh) [inline pkg/expr/vars_test.go]: Same as for Example_set_kvv

File: pkg/expr/vars_test.go [hunk 13]
```diff
@@ -8,6 +8,69 @@ import (
...
+			"foo": &Vars{value: map[string]TypedValue{
+				"k1": &String{value: "v1"},
+			}},
+			"bar": &Vars{value: map[string]TypedValue{
+				"k2": Must(NewString("v2")),
+			}},
+		}
+	)
+
+	eval(`merge(vars, bar)`, p)
```
Comment (darh) [inline pkg/expr/vars_test.go]: Missing output?

File: pkg/expr/vars.go [hunk 14]
```diff
@@ -58,8 +58,26 @@ func (t Vars) ResolveTypes(res func(typ string) Type) (err error) {
...
+		})
+	}
+
+	return vars, nil
+}
+
+// Deprecated: MustMerge combines the given Vars(es) into Vars
+// It will be now Merge(nn ...Iterator) (TypedValue, error)
+// Which does the same except it will return TypedValue and error(if necessary)
+func (t *Vars) MustMerge(nn ...Iterator) *Vars {
```
Comment (darh) [inline pkg/expr/vars.go:79]: Why is it deprecated? :)
Let's call it a shorthand function.

**Review by darh — CHANGES_REQUESTED** [review]
Almost there...

File: pkg/expr/expr_types_test.go (hunk 10 above)
Comment (vicpatel) [reply pkg/expr/expr_types_test.go]: Sure

**Pull request merged**